include/vcpkg/base/parallel-algorithms.h

```diff
@@ -1,85 +1,163 @@
 #pragma once
+#if defined(_WIN32)
+#include <vcpkg/base/system-headers.h>
+#else // ^^^ _WIN32 / !_WIN32 vvv
+#include <vcpkg/base/optional.h>
 
+#include <system_error>
+#include <thread>
+#include <type_traits>
+#include <vector>
+#endif // ^^^ _WIN32
 #include <vcpkg/base/system.h>
 
+#include <algorithm>
 #include <atomic>
-#include <future>
-#include <vector>
 
 namespace vcpkg
 {
     template<class F>
-    inline void execute_in_parallel(size_t work_count, F&& work) noexcept
+    struct WorkCallbackContext
     {
-        const size_t thread_count = static_cast<size_t>(get_concurrency());
-        const size_t num_threads = std::max(static_cast<size_t>(1), std::min(thread_count, work_count));
+        F work;
+        size_t work_count;
+        std::atomic<size_t> next_offset;
 
-        std::vector<std::future<void>> workers;
-        workers.reserve(num_threads - 1);
+        WorkCallbackContext(F init_f, size_t work_count) : work(init_f), work_count(work_count), next_offset(0) { }
 
-        for (size_t i = 0; i < num_threads - 1; ++i)
+        void run()
         {
-            workers.emplace_back(std::async(std::launch::async | std::launch::deferred, [&work]() { work(); }));
+            auto offset = next_offset.load(std::memory_order_relaxed);
+            while (offset < work_count)
+            {
+                if (!next_offset.compare_exchange_weak(offset, offset + 1, std::memory_order_relaxed))
```

Comment:
Could this use `fetch_add` instead, since we can assume that `work_count + threads < SIZE_MAX`?

Reply:
I'm not sure we *can* make that assumption. (`fetch_add` was @Thomas1664 's initial solution and I requested this form to avoid overflow)

I suppose we can assume this is unlikely and add other edge cases for it though I'm not sure that would be better... stand by

Reply:
> Could this use `fetch_add` instead, since we can assume that `work_count + threads < SIZE_MAX`?

`work_count` could reach `SIZE_MAX`. I remember that some PRs that used smaller types in context of the install plan were rejected.